Has anyone tried to build native micronaut with cassandra feature?

Has anyone tried to build native micronaut with cassandra feature? Please help. When I build simple micronaut cassandra application with graalvm below exception thrown:


`
applicationType: default
defaultPackage: com.example
testFramework: junit
sourceLanguage: java
buildTool: gradle
features: [app-name, cassandra, docker, file-watch, graalvm, gradle, http-client, java, java-application, junit, logback, netty-server, readme, shade, yaml]
`

`native-image --no-server -cp build/libs/demo-0.1-all.jar`

`Fatal error: java.lang.NoClassDefFoundError: Lorg/apache/ivy/Ivy;
    at java.lang.Class.getDeclaredFields0(Native Method)
    at java.lang.Class.privateGetDeclaredFields(Class.java:2583)
    at java.lang.Class.getDeclaredField(Class.java:2068)
    at com.oracle.graal.pointsto.infrastructure.OriginalFieldProvider.getJavaField(OriginalFieldProvider.java:43)
    at com.oracle.graal.pointsto.meta.AnalysisField.getJavaField(AnalysisField.java:420)
    at com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.readUninitializedStaticValue(AnalysisConstantReflectionProvider.java:136)
    at com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.readValue(AnalysisConstantReflectionProvider.java:87)
    at com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.readFieldValue(AnalysisConstantReflectionProvider.java:77)
    at com.oracle.graal.pointsto.ObjectScanner.scanField(ObjectScanner.java:175)
    at com.oracle.graal.pointsto.ObjectScanner$1.run(ObjectScanner.java:103)
    at com.oracle.graal.pointsto.util.CompletionExecutor.lambda$execute$0(CompletionExecutor.java:171)
    at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
    at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
    at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
    at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
    at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: java.lang.ClassNotFoundException: org.apache.ivy.Ivy
    at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
    ... 16 more`

--- a/cassandra/src/main/resources/META-INF/native-image/io.micronaut.cassandra/cassandra/jni-config.json
+++ b/cassandra/src/main/resources/META-INF/native-image/io.micronaut.cassandra/cassandra/jni-config.json
@@ -1,0 +1,6 @@
+[
+{
+  "name":"java.lang.UnsatisfiedLinkError",
+  "methods":[{"name":"<init>","parameterTypes":["java.lang.String"] }]
+}
+]

--- a/cassandra/src/main/resources/META-INF/native-image/io.micronaut.cassandra/cassandra/native-image.properties
+++ b/cassandra/src/main/resources/META-INF/native-image/io.micronaut.cassandra/cassandra/native-image.properties
@@ -1,0 +1,17 @@
+#
+# Copyright 2017-2021 original authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+Args = --initialize-at-run-time=io.micronaut.configuration.cassandra.health.$CassandraHealthIndicatorDefinition

--- a/cassandra/src/main/resources/META-INF/native-image/io.micronaut.cassandra/cassandra/proxy-config.json
+++ b/cassandra/src/main/resources/META-INF/native-image/io.micronaut.cassandra/cassandra/proxy-config.json
@@ -1,0 +1,3 @@
+[
+  ["java.lang.reflect.TypeVariable"]
+]

--- a/cassandra/src/main/resources/META-INF/native-image/io.micronaut.cassandra/cassandra/reflect-config.json
+++ b/cassandra/src/main/resources/META-INF/native-image/io.micronaut.cassandra/cassandra/reflect-config.json
@@ -1,0 +1,138 @@
+[
+  {
+    "name":"com.datastax.oss.driver.internal.core.addresstranslation.PassThroughAddressTranslator",
+    "methods":[{"name":"<init>","parameterTypes":["com.datastax.oss.driver.api.core.context.DriverContext"] }]
+  },
+  {
+    "name":"com.datastax.oss.driver.internal.core.channel.ChannelFactory$1"
+  },
+  {
+    "name":"com.datastax.oss.driver.internal.core.channel.ConnectInitHandler",
+    "methods":[{"name":"connect","parameterTypes":["io.netty.channel.ChannelHandlerContext","java.net.SocketAddress","java.net.SocketAddress","io.netty.channel.ChannelPromise"] }]
+  },
+  {
+    "name":"com.datastax.oss.driver.internal.core.channel.HeartbeatHandler"
+  },
+  {
+    "name":"com.datastax.oss.driver.internal.core.channel.InFlightHandler",
+    "methods":[
+      {"name":"channelActive","parameterTypes":["io.netty.channel.ChannelHandlerContext"] },
+      {"name":"channelInactive","parameterTypes":["io.netty.channel.ChannelHandlerContext"] },
+      {"name":"channelRead","parameterTypes":["io.netty.channel.ChannelHandlerContext","java.lang.Object"] },
+      {"name":"exceptionCaught","parameterTypes":["io.netty.channel.ChannelHandlerContext","java.lang.Throwable"] },
+      {"name":"userEventTriggered","parameterTypes":["io.netty.channel.ChannelHandlerContext","java.lang.Object"] },
+      {"name":"write","parameterTypes":["io.netty.channel.ChannelHandlerContext","java.lang.Object","io.netty.channel.ChannelPromise"] }
+    ]
+  },
+  {
+    "name":"com.datastax.oss.driver.internal.core.channel.ProtocolInitHandler",
+    "methods":[{"name":"channelActive","parameterTypes":["io.netty.channel.ChannelHandlerContext"] }]
+  },
+  {
+    "name":"com.datastax.oss.driver.internal.core.connection.ExponentialReconnectionPolicy",
+    "methods":[{"name":"<init>","parameterTypes":["com.datastax.oss.driver.api.core.context.DriverContext"] }]
+  },
+  {
+    "name":"com.datastax.oss.driver.internal.core.loadbalancing.DefaultLoadBalancingPolicy",
+    "methods":[{"name":"<init>","parameterTypes":["com.datastax.oss.driver.api.core.context.DriverContext","java.lang.String"] }]
+  },
+  {
+    "name":"com.datastax.oss.driver.internal.core.metadata.NoopNodeStateListener",
+    "methods":[{"name":"<init>","parameterTypes":["com.datastax.oss.driver.api.core.context.DriverContext"] }]
+  },
+  {
+    "name":"com.datastax.oss.driver.internal.core.metadata.schema.NoopSchemaChangeListener",
+    "methods":[{"name":"<init>","parameterTypes":["com.datastax.oss.driver.api.core.context.DriverContext"] }]
+  },
+  {
+    "name":"com.datastax.oss.driver.internal.core.metrics.DropwizardMetricsFactory",
+    "methods":[{"name":"<init>","parameterTypes":["com.datastax.oss.driver.api.core.context.DriverContext"] }]
+  },
+  {
+    "name":"com.datastax.oss.driver.internal.core.protocol.FrameDecoder"
+  },
+  {
+    "name":"com.datastax.oss.driver.internal.core.protocol.FrameEncoder"
+  },
+  {
+    "name":"com.datastax.oss.driver.internal.core.retry.DefaultRetryPolicy",
+    "methods":[{"name":"<init>","parameterTypes":["com.datastax.oss.driver.api.core.context.DriverContext","java.lang.String"] }]
+  },
+  {
+    "name":"com.datastax.oss.driver.internal.core.session.throttling.PassThroughRequestThrottler",
+    "methods":[{"name":"<init>","parameterTypes":["com.datastax.oss.driver.api.core.context.DriverContext"] }]
+  },
+  {
+    "name":"com.datastax.oss.driver.internal.core.specex.NoSpeculativeExecutionPolicy",
+    "methods":[{"name":"<init>","parameterTypes":["com.datastax.oss.driver.api.core.context.DriverContext","java.lang.String"] }]
+  },
+  {
+    "name":"com.datastax.oss.driver.internal.core.time.AtomicTimestampGenerator",
+    "methods":[{"name":"<init>","parameterTypes":["com.datastax.oss.driver.api.core.context.DriverContext"] }]
+  },
+  {
+    "name":"com.datastax.oss.driver.internal.core.tracker.NoopRequestTracker",
+    "methods":[{"name":"<init>","parameterTypes":["com.datastax.oss.driver.api.core.context.DriverContext"] }]
+  },
+  {
+    "name":"com.datastax.oss.driver.shaded.guava.common.reflect.Types$TypeVariableImpl",
+    "allPublicMethods":true
+  },
+  {
+    "name":"com.datastax.oss.driver.shaded.guava.common.util.concurrent.AbstractFuture",
+    "fields":[
+      {"name":"listeners", "allowUnsafeAccess":true},
+      {"name":"value", "allowUnsafeAccess":true},
+      {"name":"waiters", "allowUnsafeAccess":true}
+    ]
+  },
+  {
+    "name":"com.datastax.oss.driver.shaded.guava.common.util.concurrent.AbstractFuture$Waiter",
+    "fields":[
+      {"name":"next", "allowUnsafeAccess":true},
+      {"name":"thread", "allowUnsafeAccess":true}
+    ]
+  },
+  {
+    "name":"com.esri.core.geometry.ogc.OGCGeometry"
+  },
+  {
+    "name":"com.fasterxml.jackson.core.JsonParser"
+  },
+  {
+    "name":"com.kenai.jffi.Invoker",
+    "methods":[
+      {"name":"invokeI6","parameterTypes":["com.kenai.jffi.CallContext","long","int","int","int","int","int","int"] },
+      {"name":"invokeL6","parameterTypes":["com.kenai.jffi.CallContext","long","long","long","long","long","long","long"] },
+      {"name":"invokeN6","parameterTypes":["com.kenai.jffi.CallContext","long","long","long","long","long","long","long"] }
+    ]
+  },
+  {
+    "name":"com.kenai.jffi.Version",
+    "fields":[
+      {"name":"MAJOR"},
+      {"name":"MICRO"},
+      {"name":"MINOR"}
+    ]
+  },
+  {
+    "name":"com.kenai.jffi.internal.StubLoader",
+    "methods":[{"name":"isLoaded","parameterTypes":[] }]
+  },
+  {
+    "name":"org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversal"
+  },
+  {
+    "name":"org.apache.tinkerpop.gremlin.structure.Graph"
+  },
+  {
+    "name":"org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerGraph",
+    "methods":[{"name":"open","parameterTypes":["org.apache.commons.configuration.Configuration"] }]
+  },
+  {
+    "name":"org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerIoRegistryV3d0"
+  },
+  {
+    "name":"org.reactivestreams.Publisher"
+  }
+]

--- a/cassandra/src/main/resources/META-INF/native-image/io.micronaut.cassandra/cassandra/resource-config.json
+++ b/cassandra/src/main/resources/META-INF/native-image/io.micronaut.cassandra/cassandra/resource-config.json
@@ -1,0 +1,9 @@
+{
+  "resources":{
+  "includes":[
+    {"pattern":"\\Qcom/datastax/oss/driver/Driver.properties\\E"},
+    {"pattern":"\\Qjni/x86_64-Linux/libjffi-1.2.so\\E"}, 
+    {"pattern":"\\Qreference.conf\\E"}
+  ]},
+  "bundles":[]
+}

--- a/gradle.properties
+++ b/gradle.properties
@@ -1,4 +1,4 @@
-projectVersion=3.1.2-SNAPSHOT
+projectVersion=3.2.0-SNAPSHOT
 micronautDocsVersion=1.0.25
 micronautVersion=2.1.3
 micronautTestVersion=2.3.2

--- a/src/main/docs/guide/graalvm.adoc
+++ b/src/main/docs/guide/graalvm.adoc
@@ -1,0 +1,6 @@
+
+Micronaut Cassandra is compatible with https://www.graalvm.org/[GraalVM] so it is possible
+to create native images without any additional configuration.
+
+NOTE: See the section on https://docs.micronaut.io/latest/guide/index.html#graal[GraalVM] in the user guide for more
+information.

--- a/src/main/docs/guide/toc.yml
+++ b/src/main/docs/guide/toc.yml
@@ -3,4 +3,5 @@ releaseHistory: Release History
 setup: Setting up Cassandra
 health: Health Checks
 additional: Additional Notes
+graalvm: GraalVM support
 repository: Repository

